contributors/design-proposals/azure/cross-resource-group-nodes.md

```diff
@@ -0,0 +1,124 @@
+# Cross resource group nodes
+
+- **Authors**: Pengfei Ni (@feiskyer)
+- **Last Updated**: 2018-08-06
+- **Status**: Draft
+
+This proposal aims to add cross resource group (RG) nodes and on-prem nodes support to Azure cloud provider.
+
+## Introduction
+
+Today, Azure cloud provider only supports nodes from specified RG (which is set in cloud provider configure file). For nodes in different RG, Azure cloud provider reports `InstanceNotFound` error and thus they would be removed by controller manager. So as on-prem nodes.
+
+With managed cluster (limited access to nodes) users may need nodes where they can customize in ways that are not possible in a managed service. So this document proposes to support joining arbitrary nodes to the cluster and make required changes for both Azure cloud provider and provision setups, which include:
+
+- Provision tools should setup kubelet with required labels (e.g. via `--node-labels`)
+- Azure cloud provider would fetch RG from those labels and then get node information based on that
+
+Note that provisioning Kubernetes cluster, setting up networking and provisioning new nodes are out of this proposal scope. Those could be done by external provisioning tools (e.g. acs-engine).
+
+## Assumptions
+
+While news nodes (either from different RGs or on-prem) would be supported in this proposal, not all features would be supported for them. For example, AzureDisk will not work for on-prem nodes.
+
+This proposal makes following assumptions for those new nodes:
+
+- Nodes are in same region and set with required labels (as clarified in the following design part)
+- Nodes will not be part of the load balancer managed by cloud provider
+- Both node and container networking have setup properly
+- AzureDisk is supported for Azure cross-RG nodes, but not for on-prem nodes
+
+Besides, feature gate [ServiceNodeExclusion](https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go#L174) should also be enabled for Kubernetes cluster.
+
+## Design
+
+Instance metadata is a general way to fetch node information for Azure, but it doesn't work if cloud-controller-manager is used (`kubelet --cloud-provider=external`). So it won't be used in this proposal. Instead, following two labels are proposed for providing required information:
```

Comment:
use of the external cloud provider typically involves a controller that pulls information from metadata and sets it on Node objects. Using a source of truth outside the individual kubelets is preferable for setting topology information on Node objects, since it doesn't require trusting all nodes to categorize themselves appropriately.